apple/internal/apple_library.bzl

```diff
@@ -0,0 +1,309 @@
+#import <Foundation/Foundation.h>
+#if __has_include(<UIKit/UIKit.h>)
+#import <UIKit/UIKit.h>
+#endif
+#endif
+
+"""
+    for hdr in hdrs:
+        if hdr.extension == "h":
+            content += "#import \"%s\"\n" % (hdr.path)
+
+    return content
+
+def _module_map_impl(ctx):
+    outputs = []
+
+    hdrs = ctx.files.hdrs
+    textual_hdrs = ctx.files.textual_hdrs
+    outputs.extend(hdrs)
+    outputs.extend(textual_hdrs)
+
+    # Find Swift generated header
+    swift_generated_header = None
+    for dep in ctx.attr.deps:
+        if CcInfo in dep:
+            objc_headers = dep[CcInfo].compilation_context.headers.to_list()
+        else:
+            objc_headers = []
+        for hdr in objc_headers:
+            if hdr.owner == dep.label:
+                swift_generated_header = hdr
+                outputs.append(swift_generated_header)
+
+    # Write the module map content
+    if swift_generated_header:
+        umbrella_header_path = ctx.attr.module_name + ".h"
+        umbrella_header = ctx.actions.declare_file(umbrella_header_path)
+        outputs.append(umbrella_header)
+        ctx.actions.write(
+            content = _umbrella_header_content(hdrs),
+            output = umbrella_header,
+        )
+        outputs.append(umbrella_header)
+
+    module_map = ctx.actions.declare_file(ctx.attr.name + "-module.modulemap")
+    outputs.append(module_map)
+
+    ctx.actions.write(
+        content = _module_map_content(
+            module_name = ctx.attr.module_name,
+            hdrs = hdrs,
+            textual_hdrs = textual_hdrs,
+            swift_generated_header = swift_generated_header,
+            module_map_path = module_map.path,
+        ),
+        output = module_map,
+    )
+
+    objc_provider = apple_common.new_objc_provider()
+
+    compilation_context = cc_common.create_compilation_context(
+        headers = depset(outputs),
+    )
+    cc_info = CcInfo(
+        compilation_context = compilation_context,
+    )
+
+    return [
+        DefaultInfo(
+            files = depset([module_map]),
+        ),
+        objc_provider,
+        cc_info,
+    ]
+
+_module_map = rule(
+    attrs = {
+        "module_name": attr.string(
+            mandatory = True,
+            doc = "The name of the module.",
+        ),
+        "hdrs": attr.label_list(
+            allow_files = _HEADERS_FILE_TYPES,
+            doc = """\
+The list of C, C++, Objective-C, and Objective-C++ header files used to
+construct the module map.
+""",
+        ),
+        "textual_hdrs": attr.label_list(
+            allow_files = _HEADERS_FILE_TYPES,
+            doc = """\
+The list of C, C++, Objective-C, and Objective-C++ header files used to
+construct the module map. Unlike hdrs, these will be declared as 'textual
+header' in the module map.
+""",
+        ),
+        "deps": attr.label_list(
+            providers = [SwiftInfo],
+            doc = """\
+The list of swift_library targets.  A `${module_name}.Swift` submodule will be
```

Comment:
If we could change this to not physically depend on the swift_library; we could work around https://github.com/buildbuddy-io/rules_xcodeproj/blob/main/docs/faq.md#why-do-some-of-my-swift_librarys-compile-twice-in-bwx-mode. We could pass in any info needed for the generated header path.

Not a blocker, just a shower thought.

Reply:
I think I've thought about this when first implemented the rule in the past, but don't remember the details. I can take a look again once this is merged.